File: src/main/host/syscall/handler/uio.rs
use linux_api::errno::Errno;
use shadow_shim_helper_rs::syscall_types::ForeignPtr;

use crate::cshadow as c;
use crate::host::descriptor::socket::{RecvmsgArgs, RecvmsgReturn, SendmsgArgs, Socket};
use crate::host::descriptor::{CompatFile, File, FileState, FileStatus};
use crate::host::syscall::handler::{SyscallContext, SyscallHandler};
use crate::host::syscall::io::{self, IoVec};
use crate::host::syscall::types::{ForeignArrayPtr, SyscallError};
use crate::utility::callback_queue::CallbackQueue;

impl SyscallHandler {
    log_syscall!(
        readv,
        /* rv */ libc::ssize_t,
        /* fd */ std::ffi::c_int,
        /* iov */ *const libc::iovec,
        /* iovcnt */ std::ffi::c_int,
    );
    pub fn readv(
        ctx: &mut SyscallContext,
        fd: std::ffi::c_int,
        iov_ptr: ForeignPtr<libc::iovec>,
        iov_count: std::ffi::c_int,
    ) -> Result<libc::ssize_t, SyscallError> {
        // if we were previously blocked, get the active file from the last syscall handler
        // invocation since it may no longer exist in the descriptor table
        let file = ctx
            .objs
            .thread
            .syscall_condition()
            // if this was for a C descriptor, then there won't be an active file object
            .and_then(|x| x.active_file().cloned());

        let file = match file {
            // we were previously blocked, so re-use the file from the previous syscall invocation
            Some(x) => x,
            // get the file from the descriptor table, or return early if it doesn't exist
            None => {
                let desc_table = ctx.objs.thread.descriptor_table_borrow(ctx.objs.host);
                match Self::get_descriptor(&desc_table, fd)?.file() {
                    CompatFile::New(file) => file.clone(),
                    // if it's a legacy file, use the C syscall handler instead
                    CompatFile::Legacy(_) => {
                        drop(desc_table);
                        return Self::legacy_syscall(c::syscallhandler_readv, ctx);
                    }
                }
            }
        };

        let iov_count = iov_count.try_into().or(Err(Errno::EINVAL))?;

        let iovs = {
            let mem = ctx.objs.process.memory_borrow_mut();
            io::read_iovecs(&mem, iov_ptr, iov_count)?
        };
        assert_eq!(iovs.len(), iov_count);

        let mut result = Self::readv_helper(ctx, file.inner_file(), &iovs, None, 0);

        // if the syscall will block, keep the file open until the syscall restarts
        if let Some(err) = result.as_mut().err() {
            if let Some(cond) = err.blocked_condition() {
                cond.set_active_file(file);
            }
        }

        let bytes_read = result?;
        Ok(bytes_read)
    }

    log_syscall!(
        preadv,
        /* rv */ libc::ssize_t,
        /* fd */ std::ffi::c_int,
        /* iov */ *const libc::iovec,
        /* iovcnt */ std::ffi::c_int,
        /* pos_l */ libc::c_ulong,
        /* pos_h */ libc::c_ulong,
    );
    pub fn preadv(
        ctx: &mut SyscallContext,
        fd: std::ffi::c_int,
        iov_ptr: ForeignPtr<libc::iovec>,
        iov_count: std::ffi::c_int,
        offset_l: libc::c_ulong,
        _offset_h: libc::c_ulong,
    ) -> Result<libc::ssize_t, SyscallError> {
        // on Linux x86-64, an `unsigned long` is 64 bits, so we can ignore `offset_h`
        static_assertions::assert_eq_size!(libc::c_ulong, libc::off_t);
        let offset = offset_l as libc::off_t;

        // if we were previously blocked, get the active file from the last syscall handler
        // invocation since it may no longer exist in the descriptor table
        let file = ctx
            .objs
            .thread
            .syscall_condition()
            // if this was for a C descriptor, then there won't be an active file object
            .and_then(|x| x.active_file().cloned());

        let file = match file {
            // we were previously blocked, so re-use the file from the previous syscall invocation
            Some(x) => x,
            // get the file from the descriptor table, or return early if it doesn't exist
            None => {
                let desc_table = ctx.objs.thread.descriptor_table_borrow(ctx.objs.host);
                match Self::get_descriptor(&desc_table, fd)?.file() {
                    CompatFile::New(file) => file.clone(),
                    // if it's a legacy file, use the C syscall handler instead
                    CompatFile::Legacy(_) => {
                        drop(desc_table);
                        return Self::legacy_syscall(c::syscallhandler_preadv, ctx);
                    }
                }
            }
        };

        // make sure the offset is not negative
        if offset < 0 {
            return Err(Errno::EINVAL.into());
        }

        let iov_count = iov_count.try_into().or(Err(Errno::EINVAL))?;

        let iovs = {
            let mem = ctx.objs.process.memory_borrow_mut();
            io::read_iovecs(&mem, iov_ptr, iov_count)?
        };
        assert_eq!(iovs.len(), iov_count);

        let mut result = Self::readv_helper(ctx, file.inner_file(), &iovs, Some(offset), 0);

        // if the syscall will block, keep the file open until the syscall restarts
        if let Some(err) = result.as_mut().err() {
            if let Some(cond) = err.blocked_condition() {
                cond.set_active_file(file);
            }
        }

        let bytes_read = result?;
        Ok(bytes_read)
    }

    log_syscall!(
        preadv2,
        /* rv */ libc::ssize_t,
        /* fd */ std::ffi::c_int,
        /* iov */ *const libc::iovec,
        /* iovcnt */ std::ffi::c_int,
        /* pos_l */ libc::c_ulong,
        /* pos_h */ libc::c_ulong,
        /* flags */ std::ffi::c_int,
    );
    pub fn preadv2(
        ctx: &mut SyscallContext,
        fd: std::ffi::c_int,
        iov_ptr: ForeignPtr<libc::iovec>,
        iov_count: std::ffi::c_int,
        offset_l: libc::c_ulong,
        _offset_h: libc::c_ulong,
        flags: std::ffi::c_int,
    ) -> Result<libc::ssize_t, SyscallError> {
        // on Linux x86-64, an `unsigned long` is 64 bits, so we can ignore `offset_h`
        static_assertions::assert_eq_size!(libc::c_ulong, libc::off_t);
        let offset = offset_l as libc::off_t;

        // if we were previously blocked, get the active file from the last syscall handler
        // invocation since it may no longer exist in the descriptor table
        let file = ctx
            .objs
            .thread
            .syscall_condition()
            // if this was for a C descriptor, then there won't be an active file object
            .and_then(|x| x.active_file().cloned());

        let file = match file {
            // we were previously blocked, so re-use the file from the previous syscall invocation
            Some(x) => x,
            // get the file from the descriptor table, or return early if it doesn't exist
            None => {
                let desc_table = ctx.objs.thread.descriptor_table_borrow(ctx.objs.host);
                match Self::get_descriptor(&desc_table, fd)?.file() {
                    CompatFile::New(file) => file.clone(),
                    // if it's a legacy file, use the C syscall handler instead
                    CompatFile::Legacy(_) => {
                        drop(desc_table);
                        return Self::legacy_syscall(c::syscallhandler_preadv2, ctx);
                    }
                }
            }
        };

        // readv(2): "Unlike preadv() and pwritev(), if the offset argument is -1, then the current
        // file offset is used and updated."
        let offset = (offset != -1).then_some(offset);

        // if the offset is set, make sure it's not negative
        if let Some(offset) = offset {
            if offset < 0 {
                return Err(Errno::EINVAL.into());
            }
        }

        let iov_count = iov_count.try_into().or(Err(Errno::EINVAL))?;

        let iovs = {
            let mem = ctx.objs.process.memory_borrow_mut();
            io::read_iovecs(&mem, iov_ptr, iov_count)?
        };
        assert_eq!(iovs.len(), iov_count);

        let mut result = Self::readv_helper(ctx, file.inner_file(), &iovs, offset, flags);

        // if the syscall will block, keep the file open until the syscall restarts
        if let Some(err) = result.as_mut().err() {
            if let Some(cond) = err.blocked_condition() {
                cond.set_active_file(file);
            }
        }

        let bytes_read = result?;
        Ok(bytes_read)
    }

    pub fn readv_helper(
        ctx: &mut SyscallContext,
        file: &File,
        iovs: &[IoVec],
        offset: Option<libc::off_t>,
        flags: std::ffi::c_int,
    ) -> Result<libc::ssize_t, SyscallError> {
        let mut mem = ctx.objs.process.memory_borrow_mut();

        // if it's a socket, call recvmsg_helper() instead
        if let File::Socket(ref socket) = file {
            if offset.is_some() {
                // sockets don't support offsets
                return Err(Errno::ESPIPE.into());
            }

            // experimentally, it seems that read() calls on sockets with 0-length buffers will
            // always return 0, even if there would otherwise be an EWOULDBOCK from a recv() call
            // (see the `test_zero_len_buf_read_and_recv` and `test_zero_len_msg_read_and_recv`
            // send/recv tests for examples)
            if iovs.iter().map(|x| x.len).sum::<usize>() == 0 {
                return Ok(0);
            }

            let args = RecvmsgArgs {
                iovs,
                control_ptr: ForeignArrayPtr::new(ForeignPtr::null(), 0),
                flags: 0,
            };

            // call the socket's recvmsg(), and run any resulting events
            let RecvmsgReturn { return_val, .. } =
                crate::utility::legacy_callback_queue::with_global_cb_queue(|| {
                    CallbackQueue::queue_and_run(|cb_queue| {
                        Socket::recvmsg(socket, args, &mut mem, cb_queue)
                    })
                })?;

            return Ok(return_val);
        }

        let file_status = file.borrow().status();

        let result =
            // call the file's read(), and run any resulting events
            crate::utility::legacy_callback_queue::with_global_cb_queue(|| {CallbackQueue::queue_and_run(|cb_queue| {
                file.borrow_mut().readv(
                    iovs,
                    offset,
                    flags,
                    &mut mem,
                    cb_queue,
                )
            })});

        // if the syscall would block and it's a blocking descriptor
        if result == Err(Errno::EWOULDBLOCK.into()) && !file_status.contains(FileStatus::NONBLOCK) {
            return Err(SyscallError::new_blocked_on_file(
                file.clone(),
                FileState::READABLE,
                file.borrow().supports_sa_restart(),
            ));
        }

        result
    }

    log_syscall!(
        writev,
        /* rv */ libc::ssize_t,
        /* fd */ std::ffi::c_int,
        /* iov */ *const libc::iovec,
        /* iovcnt */ std::ffi::c_int,
    );
    pub fn writev(
        ctx: &mut SyscallContext,
        fd: std::ffi::c_int,
        iov_ptr: ForeignPtr<libc::iovec>,
        iov_count: std::ffi::c_int,
    ) -> Result<libc::ssize_t, SyscallError> {
        // if we were previously blocked, get the active file from the last syscall handler
        // invocation since it may no longer exist in the descriptor table
        let file = ctx
            .objs
            .thread
            .syscall_condition()
            // if this was for a C descriptor, then there won't be an active file object
            .and_then(|x| x.active_file().cloned());

        let file = match file {
            // we were previously blocked, so re-use the file from the previous syscall invocation
            Some(x) => x,
            // get the file from the descriptor table, or return early if it doesn't exist
            None => {
                let desc_table = ctx.objs.thread.descriptor_table_borrow(ctx.objs.host);
                match Self::get_descriptor(&desc_table, fd)?.file() {
                    CompatFile::New(file) => file.clone(),
                    // if it's a legacy file, use the C syscall handler instead
                    CompatFile::Legacy(_) => {
                        drop(desc_table);
                        return Self::legacy_syscall(c::syscallhandler_writev, ctx);
                    }
                }
            }
        };

        let iov_count = iov_count.try_into().or(Err(Errno::EINVAL))?;

        let iovs = {
            let mem = ctx.objs.process.memory_borrow_mut();
            io::read_iovecs(&mem, iov_ptr, iov_count)?
        };
        assert_eq!(iovs.len(), iov_count);

        let mut result = Self::writev_helper(ctx, file.inner_file(), &iovs, None, 0);

        // if the syscall will block, keep the file open until the syscall restarts
        if let Some(err) = result.as_mut().err() {
            if let Some(cond) = err.blocked_condition() {
                cond.set_active_file(file);
            }
        }

        let bytes_written = result?;
        Ok(bytes_written)
    }

    log_syscall!(
        pwritev,
        /* rv */ libc::ssize_t,
        /* fd */ std::ffi::c_int,
        /* iov */ *const libc::iovec,
        /* iovcnt */ std::ffi::c_int,
        /* pos_l */ libc::c_ulong,
        /* pos_h */ libc::c_ulong,
    );
    pub fn pwritev(
        ctx: &mut SyscallContext,
        fd: std::ffi::c_int,
        iov_ptr: ForeignPtr<libc::iovec>,
        iov_count: std::ffi::c_int,
        offset_l: libc::c_ulong,
        _offset_h: libc::c_ulong,
    ) -> Result<libc::ssize_t, SyscallError> {
        // on Linux x86-64, an `unsigned long` is 64 bits, so we can ignore `offset_h`
        static_assertions::assert_eq_size!(libc::c_ulong, libc::off_t);
        let offset = offset_l as libc::off_t;

        // if we were previously blocked, get the active file from the last syscall handler
        // invocation since it may no longer exist in the descriptor table
        let file = ctx
            .objs
            .thread
            .syscall_condition()
            // if this was for a C descriptor, then there won't be an active file object
            .and_then(|x| x.active_file().cloned());

        let file = match file {
            // we were previously blocked, so re-use the file from the previous syscall invocation
            Some(x) => x,
            // get the file from the descriptor table, or return early if it doesn't exist
            None => {
                let desc_table = ctx.objs.thread.descriptor_table_borrow(ctx.objs.host);
                match Self::get_descriptor(&desc_table, fd)?.file() {
                    CompatFile::New(file) => file.clone(),
                    // if it's a legacy file, use the C syscall handler instead
                    CompatFile::Legacy(_) => {
                        drop(desc_table);
                        return Self::legacy_syscall(c::syscallhandler_pwritev, ctx);
                    }
                }
            }
        };

        // make sure the offset is not negative
        if offset < 0 {
            return Err(Errno::EINVAL.into());
        }

        let iov_count = iov_count.try_into().or(Err(Errno::EINVAL))?;

        let iovs = {
            let mem = ctx.objs.process.memory_borrow_mut();
            io::read_iovecs(&mem, iov_ptr, iov_count)?
        };
        assert_eq!(iovs.len(), iov_count);

        let mut result = Self::writev_helper(ctx, file.inner_file(), &iovs, Some(offset), 0);

        // if the syscall will block, keep the file open until the syscall restarts
        if let Some(err) = result.as_mut().err() {
            if let Some(cond) = err.blocked_condition() {
                cond.set_active_file(file);
            }
        }

        let bytes_written = result?;
        Ok(bytes_written)
    }

    log_syscall!(
        pwritev2,
        /* rv */ libc::ssize_t,
        /* fd */ std::ffi::c_int,
        /* iov */ *const libc::iovec,
        /* iovcnt */ std::ffi::c_int,
        /* pos_l */ libc::c_ulong,
        /* pos_h */ libc::c_ulong,
        /* flags */ std::ffi::c_int,
    );
    pub fn pwritev2(
        ctx: &mut SyscallContext,
        fd: std::ffi::c_int,
        iov_ptr: ForeignPtr<libc::iovec>,
        iov_count: std::ffi::c_int,
        offset_l: libc::c_ulong,
        _offset_h: libc::c_ulong,
        flags: std::ffi::c_int,
    ) -> Result<libc::ssize_t, SyscallError> {
        // on Linux x86-64, an `unsigned long` is 64 bits, so we can ignore `offset_h`
        static_assertions::assert_eq_size!(libc::c_ulong, libc::off_t);
        let offset = offset_l as libc::off_t;

        // if we were previously blocked, get the active file from the last syscall handler
        // invocation since it may no longer exist in the descriptor table
        let file = ctx
            .objs
            .thread
            .syscall_condition()
            // if this was for a C descriptor, then there won't be an active file object
            .and_then(|x| x.active_file().cloned());

        let file = match file {
            // we were previously blocked, so re-use the file from the previous syscall invocation
            Some(x) => x,
            // get the file from the descriptor table, or return early if it doesn't exist
            None => {
                let desc_table = ctx.objs.thread.descriptor_table_borrow(ctx.objs.host);
                match Self::get_descriptor(&desc_table, fd)?.file() {
                    CompatFile::New(file) => file.clone(),
                    // if it's a legacy file, use the C syscall handler instead
                    CompatFile::Legacy(_) => {
                        drop(desc_table);
                        return Self::legacy_syscall(c::syscallhandler_pwritev2, ctx);
                    }
                }
            }
        };

        // readv(2): "Unlike preadv() and pwritev(), if the offset argument is -1, then the current
        // file offset is used and updated."
        let offset = (offset != -1).then_some(offset);

        // if the offset is set, make sure it's not negative
        if let Some(offset) = offset {
            if offset < 0 {
                return Err(Errno::EINVAL.into());
            }
        }

        let iov_count = iov_count.try_into().or(Err(Errno::EINVAL))?;

        let iovs = {
            let mem = ctx.objs.process.memory_borrow_mut();
            io::read_iovecs(&mem, iov_ptr, iov_count)?
        };
        assert_eq!(iovs.len(), iov_count);

        let mut result = Self::writev_helper(ctx, file.inner_file(), &iovs, offset, flags);

        // if the syscall will block, keep the file open until the syscall restarts
        if let Some(err) = result.as_mut().err() {
            if let Some(cond) = err.blocked_condition() {
                cond.set_active_file(file);
            }
        }

        let bytes_written = result?;
        Ok(bytes_written)
    }

    pub fn writev_helper(
        ctx: &mut SyscallContext,
        file: &File,
        iovs: &[IoVec],
        offset: Option<libc::off_t>,
        flags: std::ffi::c_int,
    ) -> Result<libc::ssize_t, SyscallError> {
        let mut mem = ctx.objs.process.memory_borrow_mut();
        let mut rng = ctx.objs.host.random_mut();
        let net_ns = ctx.objs.host.network_namespace_borrow();

        // if it's a socket, call sendmsg_helper() instead
        if let File::Socket(ref socket) = file {
            if offset.is_some() {
                // sockets don't support offsets
                return Err(Errno::ESPIPE.into());
            }

            let args = SendmsgArgs {
                addr: None,
                iovs,
                control_ptr: ForeignArrayPtr::new(ForeignPtr::null(), 0),
                flags: 0,
            };

            // call the socket's sendmsg(), and run any resulting events
            let bytes_written =
                crate::utility::legacy_callback_queue::with_global_cb_queue(|| {
                    CallbackQueue::queue_and_run(|cb_queue| {
                        Socket::sendmsg(socket, args, &mut mem, &net_ns, &mut *rng, cb_queue)
                    })
                })?;

            return Ok(bytes_written);
        }

        let file_status = file.borrow().status();

        let result =
            // call the file's write(), and run any resulting events
            crate::utility::legacy_callback_queue::with_global_cb_queue(|| {CallbackQueue::queue_and_run(|cb_queue| {
                file.borrow_mut().writev(
                    iovs,
                    offset,
                    flags,
                    &mut mem,
                    cb_queue,
                )
            })});

        // if the syscall would block and it's a blocking descriptor
        if result == Err(Errno::EWOULDBLOCK.into()) && !file_status.contains(FileStatus::NONBLOCK) {
            return Err(SyscallError::new_blocked_on_file(
                file.clone(),
                FileState::WRITABLE,
                file.borrow().supports_sa_restart(),
            ));
        }

        result
    }
}

File: src/main/host/syscall/handler/uio.c
/*
 * The Shadow Simulator
 * See LICENSE for licensing information
 */

#include "main/host/syscall/handler/uio.h"

#include <assert.h>
#include <errno.h>
#include <sys/syscall.h>
#include <sys/uio.h>

#include "lib/logger/logger.h"
#include "main/bindings/c/bindings.h"
#include "main/host/descriptor/descriptor.h"
#include "main/host/descriptor/regular_file.h"
#include "main/host/process.h"
#include "main/host/syscall/handler/unistd.h"
#include "main/host/syscall/protected.h"
#include "main/host/syscall/syscall_condition.h"

///////////////////////////////////////////////////////////
// Helpers
///////////////////////////////////////////////////////////

static int _syscallhandler_validateVecParams(SyscallHandler* sys, int fd, UntypedForeignPtr iovPtr,
                                             unsigned long iovlen, off_t offset,
                                             LegacyFile** desc_out, struct iovec** iov_out) {
    /* Get the descriptor. */
    LegacyFile* desc = thread_getRegisteredLegacyFile(rustsyscallhandler_getThread(sys), fd);
    if (!desc) {
        return -EBADF;
    }

    /* Validate vector length param. */
    if (iovlen == 0) {
        return 0;
    } else if (iovlen > UIO_MAXIOV) {
        return -EINVAL;
    }

    /* Make sure we have a non-null vector. */
    if (!iovPtr.val) {
        return -EFAULT;
    }

    /* We can only seek on files, otherwise its a pipe error. */
    if (legacyfile_getType(desc) != DT_FILE && offset != 0) {
        return -ESPIPE;
    }

    /* Get the vector of pointers. */
    struct iovec* iov = malloc(iovlen * sizeof(*iov));
    if (process_readPtr(rustsyscallhandler_getProcess(sys), iov, iovPtr, iovlen * sizeof(*iov)) != 0) {
        warning("Got unreadable pointer [%p..+%zu]", (void*)iovPtr.val, iovlen * sizeof(*iov));
        free(iov);
        return -EFAULT;
    }

    /* Check that all of the buf pointers are valid. */
    for (unsigned long i = 0; i < iovlen; i++) {
        UntypedForeignPtr bufPtr = (UntypedForeignPtr){.val = (uint64_t)iov[i].iov_base};
        size_t bufSize = iov[i].iov_len;

        if (!bufPtr.val && bufSize != 0) {
            debug("Invalid NULL pointer in iovec[%ld]", i);
            free(iov);
            return -EFAULT;
        }
    }

    if (desc_out) {
        *desc_out = desc;
    }
    if (iov_out) {
        *iov_out = iov;
    }
    return 0;
}

static SyscallReturn _syscallhandler_readvHelper(SyscallHandler* sys, int fd,
                                                 UntypedForeignPtr iovPtr, unsigned long iovlen,
                                                 unsigned long pos_l, unsigned long pos_h,
                                                 int flags, bool doPreadv,
                                                 bool negativeOffsetDisables) {
    /* On Linux x86-64, an `unsigned long` is 64 bits, so we can ignore `pos_h`. */
    static_assert(sizeof(unsigned long) == sizeof(off_t), "Unexpected `unsigned long` size");
    off_t offset = pos_l;

    /* If the offset is -1 for preadv2, disable the offset. */
    if (offset == -1 && negativeOffsetDisables) {
        offset = 0;
        doPreadv = false;
    }

    trace("Trying to readv from fd %d, ptr %p, size %zu, pos_l %lu, pos_h %lu, "
          "offset %ld, flags %d",
          fd, (void*)iovPtr.val, iovlen, pos_l, pos_h, offset, flags);

    if (offset < 0 && doPreadv) {
        return syscallreturn_makeDoneI64(-EINVAL);
    }

    LegacyFile* desc = NULL;
    struct iovec* iov = NULL;
    int errcode = _syscallhandler_validateVecParams(
        sys, fd, iovPtr, iovlen, offset, &desc, &iov);
    if (errcode < 0 || iovlen == 0) {
        if (iov != NULL) {
            free(iov);
        }
        return syscallreturn_makeDoneI64(errcode);
    }

    /* Some logic depends on the descriptor type. */
    LegacyFileType dType = legacyfile_getType(desc);

    ssize_t result = 0;

    /* Now we can perform the read operations. */

    /* For non-files, we only read one buffer at a time to avoid
     * unnecessary data transfer between the plugin and Shadow. */
    size_t totalBytesWritten = 0;

    for (unsigned long i = 0; i < iovlen; i++) {
        UntypedForeignPtr bufPtr = (UntypedForeignPtr){.val = (uint64_t)iov[i].iov_base};
        size_t bufSize = iov[i].iov_len;

        if (bufSize == 0) {
            /* Nothing to do if the buffer is empty. */
            continue;
        }

        switch (dType) {
            case DT_FILE: {
                off_t thisOffset = offset;

                if (doPreadv) {
                    thisOffset += totalBytesWritten;
                }

                SyscallReturn scr =
                    _syscallhandler_readHelper(sys, fd, bufPtr, bufSize, thisOffset, doPreadv);

                // if the above syscall handler created any pointers, we may
                // need to flush them before calling the syscall handler again
                result = process_flushPtrs(rustsyscallhandler_getProcess(sys));
                if (result != 0) {
                    break;
                }

                switch (scr.tag) {
                    case SYSCALL_RETURN_DONE: {
                        result = syscallreturn_done(&scr)->retval.as_i64;
                        break;
                    }
                    case SYSCALL_RETURN_BLOCK: {
                        // assume that there was no timer, and that we're blocked on this socket
                        SyscallReturnBlocked* blocked = syscallreturn_blocked(&scr);
                        syscallcondition_unref(blocked->cond);
                        result = -EWOULDBLOCK;
                        break;
                    }
                    case SYSCALL_RETURN_NATIVE: {
                        panic("recv() returned SYSCALL_NATIVE");
                    }
                }

                break;
            }
            case DT_EPOLL:
            default: {
                warning("readv() not yet implemented for descriptor type %i", (int)dType);
                result = -ENOTSUP;
                break;
            }
        }

        if (result > 0) {
            totalBytesWritten += (size_t)result;
        } else {
            break;
        }
    }
    if (result >= 0 || (result == -EWOULDBLOCK && totalBytesWritten > 0)) {
        result = totalBytesWritten;
    }

    free(iov);

    if (result == -EWOULDBLOCK && !(legacyfile_getFlags(desc) & O_NONBLOCK)) {
        /* Blocking for file io will lock up the plugin because we don't
         * yet have a way to wait on file descriptors. */
        if (dType == DT_FILE) {
            error("Indefinitely blocking a readv of vector length %lu on "
                  "file %i at offset %li",
                  iovlen, fd, offset);
        }

        /* We need to block until the descriptor is ready to write. */
        Trigger trigger =
            (Trigger){.type = TRIGGER_DESCRIPTOR, .object = desc, .status = FileState_READABLE};

        return syscallreturn_makeBlocked(
            syscallcondition_new(trigger), legacyfile_supportsSaRestart(desc));
    }

    return syscallreturn_makeDoneI64(result);
}

static SyscallReturn _syscallhandler_writevHelper(SyscallHandler* sys, int fd,
                                                  UntypedForeignPtr iovPtr, unsigned long iovlen,
                                                  unsigned long pos_l, unsigned long pos_h,
                                                  int flags, bool doPwritev,
                                                  bool negativeOffsetDisables) {
    /* On Linux x86-64, an `unsigned long` is 64 bits, so we can ignore `pos_h`. */
    static_assert(sizeof(unsigned long) == sizeof(off_t), "Unexpected `unsigned long` size");
    off_t offset = pos_l;

    /* If the offset is -1 for pwritev2, disable the offset. */
    if (offset == -1 && negativeOffsetDisables) {
        offset = 0;
        doPwritev = false;
    }

    trace("Trying to writev to fd %d, ptr %p, size %zu, pos_l %lu, pos_h %lu, "
          "offset %ld, flags %d",
          fd, (void*)iovPtr.val, iovlen, pos_l, pos_h, offset, flags);

    if (offset < 0 && doPwritev) {
        return syscallreturn_makeDoneI64(-EINVAL);
    }

    LegacyFile* desc = NULL;
    struct iovec* iov = NULL;
    int errcode = _syscallhandler_validateVecParams(
        sys, fd, iovPtr, iovlen, offset, &desc, &iov);
    if (errcode < 0 || iovlen == 0) {
        if (iov != NULL) {
            free(iov);
        }
        return syscallreturn_makeDoneI64(errcode);
    }

    /* Some logic depends on the descriptor type. */
    LegacyFileType dType = legacyfile_getType(desc);

    ssize_t result = 0;

    /* Now we can perform the write operations. */

    /* For non-files, we only read one buffer at a time to avoid
     * unnecessary data transfer between the plugin and Shadow. */
    size_t totalBytesWritten = 0;

    for (unsigned long i = 0; i < iovlen; i++) {
        UntypedForeignPtr bufPtr = (UntypedForeignPtr){.val = (uint64_t)iov[i].iov_base};
        size_t bufSize = iov[i].iov_len;

        if (bufSize == 0) {
            /* Nothing to do if the buffer is empty. */
            continue;
        }

        switch (dType) {
            case DT_FILE: {
                off_t thisOffset = offset;

                if (doPwritev) {
                    thisOffset += totalBytesWritten;
                }

                SyscallReturn scr =
                    _syscallhandler_writeHelper(sys, fd, bufPtr, bufSize, thisOffset, doPwritev);

                // if the above syscall handler created any pointers, we may
                // need to flush them before calling the syscall handler again
                result = process_flushPtrs(rustsyscallhandler_getProcess(sys));
                if (result != 0) {
                    break;
                }

                switch (scr.tag) {
                    case SYSCALL_RETURN_DONE: {
                        result = syscallreturn_done(&scr)->retval.as_i64;
                        break;
                    }
                    case SYSCALL_RETURN_BLOCK: {
                        // assume that there was no timer, and that we're blocked on this socket
                        SyscallReturnBlocked* blocked = syscallreturn_blocked(&scr);
                        syscallcondition_unref(blocked->cond);
                        result = -EWOULDBLOCK;
                        break;
                    }
                    case SYSCALL_RETURN_NATIVE: {
                        panic("send() returned SYSCALL_NATIVE");
                    }
                }

                break;
            }
            case DT_EPOLL:
            default: {
                warning("writev() not yet implemented for descriptor type %i", (int)dType);
                result = -ENOTSUP;
                break;
            }
        }

        if (result > 0) {
            totalBytesWritten += (size_t)result;
        } else {
            break;
        }
    }
    if (result >= 0 || (result == -EWOULDBLOCK && totalBytesWritten > 0)) {
        result = totalBytesWritten;
    }

    free(iov);

    if (result == -EWOULDBLOCK && !(legacyfile_getFlags(desc) & O_NONBLOCK)) {
        /* Blocking for file io will lock up the plugin because we don't
         * yet have a way to wait on file descriptors. */
        if (dType == DT_FILE) {
            error("Indefinitely blocking a writev of vector length %lu on "
                  "file %i at offset %li",
                  iovlen, fd, offset);
        }

        /* We need to block until the descriptor is ready to write. */
        Trigger trigger =
            (Trigger){.type = TRIGGER_DESCRIPTOR, .object = desc, .status = FileState_WRITABLE};

        return syscallreturn_makeBlocked(
            syscallcondition_new(trigger), legacyfile_supportsSaRestart(desc));
    }

    return syscallreturn_makeDoneI64(result);
}

///////////////////////////////////////////////////////////
// System Calls
///////////////////////////////////////////////////////////

SyscallReturn syscallhandler_readv(SyscallHandler* sys, const SysCallArgs* args) {
    return _syscallhandler_readvHelper(sys, args->args[0].as_i64, args->args[1].as_ptr,
                                       args->args[2].as_u64, 0, 0, 0, false, false);
}

SyscallReturn syscallhandler_preadv(SyscallHandler* sys, const SysCallArgs* args) {
    return _syscallhandler_readvHelper(sys, args->args[0].as_i64, args->args[1].as_ptr,
                                       args->args[2].as_u64, args->args[3].as_u64,
                                       args->args[4].as_u64, 0, true, false);
}

SyscallReturn syscallhandler_preadv2(SyscallHandler* sys, const SysCallArgs* args) {
    return _syscallhandler_readvHelper(sys, args->args[0].as_i64, args->args[1].as_ptr,
                                       args->args[2].as_u64, args->args[3].as_u64,
                                       args->args[4].as_u64, args->args[5].as_i64, true, true);
}

SyscallReturn syscallhandler_writev(SyscallHandler* sys, const SysCallArgs* args) {
    return _syscallhandler_writevHelper(sys, args->args[0].as_i64, args->args[1].as_ptr,
                                        args->args[2].as_u64, 0, 0, 0, false, false);
}

SyscallReturn syscallhandler_pwritev(SyscallHandler* sys, const SysCallArgs* args) {
    return _syscallhandler_writevHelper(sys, args->args[0].as_i64, args->args[1].as_ptr,
                                        args->args[2].as_u64, args->args[3].as_u64,
                                        args->args[4].as_u64, 0, true, false);
}

SyscallReturn syscallhandler_pwritev2(SyscallHandler* sys, const SysCallArgs* args) {
    return _syscallhandler_writevHelper(sys, args->args[0].as_i64, args->args[1].as_ptr,
                                        args->args[2].as_u64, args->args[3].as_u64,
                                        args->args[4].as_u64, args->args[5].as_i64, true, true);
}
